http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/RequestAdapter.java

```diff
@@ -21,23 +21,79 @@
 import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.http.Protocol;
 import software.amazon.awssdk.http.SdkHttpMethod;
 import software.amazon.awssdk.http.SdkHttpRequest;
+import software.amazon.awssdk.utils.StringUtils;
+import software.amazon.awssdk.utils.Validate;
+import software.amazon.awssdk.utils.http.SdkHttpUtils;
 
 @SdkInternalApi
 public final class RequestAdapter {
 
+    private static final String HOST = "Host";
+    private static final List<String> IGNORE_HEADERS = Arrays.asList(HOST);
+
+    private final Protocol protocol;
+
+    public RequestAdapter(Protocol protocol) {
+        this.protocol = Validate.paramNotNull(protocol, "protocol");
+    }
+
     public HttpRequest adapt(SdkHttpRequest sdkRequest) {
         HttpMethod method = toNettyHttpMethod(sdkRequest.method());
         HttpHeaders headers = new DefaultHttpHeaders();
-        String uri = sdkRequest.getUri().toString();
+        String uri = encodedPathAndQueryParams(sdkRequest);
         DefaultHttpRequest request = new DefaultHttpRequest(HttpVersion.HTTP_1_1, method, uri, headers);
-        sdkRequest.headers().forEach(request.headers()::add);
+        addHeadersToRequest(request, sdkRequest);
         return request;
     }
 
     private static HttpMethod toNettyHttpMethod(SdkHttpMethod method) {
         return HttpMethod.valueOf(method.name());
     }
+
+    private static String encodedPathAndQueryParams(SdkHttpRequest sdkRequest) {
+        String encodedPath = sdkRequest.encodedPath();
+        if (StringUtils.isBlank(encodedPath)) {
+            encodedPath = "/";
+        }
+
+        String encodedQueryParams = SdkHttpUtils.encodeAndFlattenQueryParameters(sdkRequest.rawQueryParameters())
+                .map(queryParams -> "?" + queryParams)
+                .orElse("");
+
+        return encodedPath + encodedQueryParams;
+    }
+
+    /**
+     * Configures the headers in the specified Netty HTTP request.
+     */
+    private void addHeadersToRequest(DefaultHttpRequest httpRequest, SdkHttpRequest request) {
+        httpRequest.headers().add(HOST, getHostHeaderValue(request));
+
+        String scheme = request.getUri().getScheme();
+        if (Protocol.HTTP2 == protocol && !StringUtils.isBlank(scheme)) {
+            httpRequest.headers().add(ExtensionHeaderNames.SCHEME.text(), scheme);
+        }
+
+        // Copy over any other headers already in our request
+        request.headers().entrySet().stream()
+                /*
+                 * Skip the Host header to avoid sending it twice, which will
+                 * interfere with some signing schemes.
+                 */
+                .filter(e -> !IGNORE_HEADERS.contains(e.getKey()))
+                .forEach(e -> e.getValue().forEach(h -> httpRequest.headers().add(e.getKey(), h)));
+    }
+
+    private String getHostHeaderValue(SdkHttpRequest request) {
+        return !SdkHttpUtils.isUsingStandardPort(request.protocol(), request.port())
```

Comment:
nit: can we flip it? It seems more readable that way.

Reply:
@zoewangg it's copied straight out of [the apache implementation](https://github.com/aws/aws-sdk-java-v2/blob/2a303333f2b349209918a857a860266bb46102c2/http-clients/apache-client/src/main/java/software/amazon/awssdk/http/apache/internal/impl/ApacheHttpRequestFactory.java#L171)